staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -129,121 +122,178 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
-		if requestInfo.APIGroup == "" {
-			gvr.Group = "core"
-		}
-
-		apiservers, err := h.findServiceableByServers(gvr)
+		apiserversi, err := h.apiserverIdentityLeases()
 		if err != nil {
-			// resource wasn't found in SV informer cache which means that resource is an aggregated API
-			// or a CR. This situation is ok to be handled by local handler.
+			klog.ErrorS(err, "no apiserver identity leases found")
 			handler.ServeHTTP(w, r)
 			return
 		}
 
-		locallyServiceable, peerEndpoints, err := h.resolveServingLocation(apiservers)
-		if err != nil {
-			gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
-			klog.ErrorS(err, "error finding serviceable-by apiservers for the requested resource", "gvr", gvr)
-			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), h.serializer, gv, w, r)
-			return
-		}
-
-		// pass request to the next handler if found the gvr locally.
-		// TODO: maintain locally serviceable GVRs somewhere so that we dont have to
-		// consult the storageversion-informed map for those
-		if locallyServiceable {
+		if h.shouldServeLocally(apiserversi, gvr) {
 			handler.ServeHTTP(w, r)
 			return
 		}
 
-		if len(peerEndpoints) == 0 {
-			klog.Errorf("gvr %v is not served by anything in this cluster", gvr)
-			handler.ServeHTTP(w, r)
+		// find servers that are capable of serving this request
+		peerEndpoint, err := h.findServiceableByServers(r.Context(), gvr, apiserversi)
+		if err != nil {
+			klog.ErrorS(err, "error fetching remote server details while proxying")
+			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), h.serializer, gvr.GroupVersion(), w, r)
 			return
 		}
 
-		// otherwise, randomly select an apiserver and proxy request to it
-		rand := rand.Intn(len(peerEndpoints))
-		destServerHostPort := peerEndpoints[rand]
-		h.proxyRequestToDestinationAPIServer(r, w, destServerHostPort)
+		h.proxyRequestToDestinationAPIServer(r, w, peerEndpoint)
+
 	})
 }
 
-func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource) (*sync.Map, error) {
-	apiserversi, ok := h.svMap.Load(gvr)
-	if !ok || apiserversi == nil {
-		return nil, fmt.Errorf("no storageVersions found for the GVR: %v", gvr)
+func (h *peerProxyHandler) populateLocalDiscoveryCache() error {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(h.loopbackClientConfig)
+	if err != nil {
+		return fmt.Errorf("error creating discovery client: %w", err)
+	}
+
+	_, resourcesByGV, _, err := discoveryClient.GroupsAndMaybeResources()
+	if err != nil {
+		return fmt.Errorf("error getting API group resources from discovery: %w", err)
+	}
+
+	h.discoveryResponseCacheLock.Lock()
+	defer h.discoveryResponseCacheLock.Unlock()
+	for gv, resources := range resourcesByGV {
+		h.localDiscoveryResponseCache[gv] = resources.APIResources
+	}
```

Comment:
It's not obvious to me why the `localDiscoveryResponseCache` isn't reset to an empty map before populating.  Is this intentional?

Reply:
afaik this is only performed once on startup so `localDiscoveryResponseCache` starts as nil.

Reply:
Thats right, it should only be invoked once in the postStartHook of apiserver when we populate the cache for the first and the only time. Thats because the set of built-in resources that are served by the apiserver shouldnt change without a server restart.

But I still added a clear() at the top, following best practice.